### PR TITLE
Adding support for the H syntax, allowing to add jitter to a cron expression

### PR DESCRIPTION
**Pull request by MYDIH**
Hi !

We use this library to schedule jobs on our infrastructure, and we really miss this feature to allow reducing the load on our servers. This PR just adds the hash syntax introduced by Jenkins. Similar to abandoned #109

I only implemented the `H` special character, no alias, to save myself the hassle of handling edge cases with `#` nth-of-day pattern.

Since this feature introduce randomness, without a seed-able source, the resulting expression would be unpredictable when serialized and parsed again. I added a dependency to [seedrandom](https://www.npmjs.com/package/seedrandom) after considering other alternatives and rulling out embedding the PRNG directly in the code using this incredibly usefull [SO](https://stackoverflow.com/a/47593316/6583306) post. Let me know if an alternative would be better.

I'm not the most knowledgeable in cron expression in general, I may have missed an edge case here and there, let me know :wink: 

I implemented some tests, and updated the README,

This library code is delightful (especially the tests), it was a pleasure implementing this !

Thx for your time

### CODE REVIEW

**Comment by harrisiirak** [comment]
Hi @MYDIH, thanks for your contribution!

I'll try to check it out as soon as possible.

Best regards

**Comment by harrisiirak** [comment]
Quick thing I noticed is that `seedrandom` is rather heavy dependency (374 kB unpacked). If possible, some alternatives could be consider.

**Comment by MYDIH** [comment]
Maybe [pure-rand](https://www.npmjs.com/package/pure-rand) ? It's a very widespread package, I ruled it out first because I thought it would be heavier due to the load of features in it (and we have very simple needs) but npm reports only 82.2kB unpacked, what do you think ?

**Comment by harrisiirak** [comment]
@MYDIH 

`pure-rand` itself looks a nice library, but its size is 2/3 of this library. Would like to keep the bundle size as tiny as possible. Also check your linked SO post and it's interesting one indeed.

> after considering other alternatives and rulling out embedding the PRNG directly in the code

Why not? Depends how good randomness we need, but some simple PRNG (for example Mulberry32) should do the job.

For example, Mulberry32 and simple hashing function:

```
function xfnv1a(str: string) {
  let h = 2166136261 >>> 0;
  for (let i = 0; i < str.length; i++) {
    h ^= str.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  return () => h >>> 0;
}

function mulberry32(seed: number) {
  return () => {
    let t = (seed += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function seededRandom(str: string) {
  const seed = xfnv1a(str)();
  return mulberry32(seed);
}

const rand = seededRandom('random seed');

console.log(Math.floor(rand() * 30));
console.log(Math.floor(rand() * 30));
console.log(Math.floor(rand() * 30));

```

Not that much code to include. If better PRNG is required, then there are alternatives.

**Comment by MYDIH** [comment]
> Why not?

I think it's just a good rule of thumbs to always use a library when needed to avoid the burden of maintaining something that's not domain centric. If it was me, I would have used a simple enough algorithm as is, but I figured it was not my call to make here.

Anyway, thanks for the heavy lifting, I'll create a utils and rerun the tests soon enough :smile:

File: src/CronExpression.ts [hunk 1]
```diff
@@ -10,6 +10,7 @@ export type CronExpressionOptions = {
   nthDayOfWeek?: number;
   expression?: string;
   strict?: boolean;
+  seed?: string;
```
[the remark this answers is not in the document]
Comment (MYDIH) [reply src/CronExpression.ts:13]: I vote for hashSeed since it would be directly linked to the feature

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Randomness related utils
```
[the remark this answers is not in the document]
Comment (MYDIH) [reply src/utils.ts:1]: In the current implementation, the seed is generated on the fly using Math.random(). This means that two instances of a cron expression will not apply the same jitter

**Comment by MYDIH** [comment]
> For some reason Node.js 19/20 failed during the first run on the last commit

Weird ... I tested with `jod`/`iron`/`hydrogen` whitout issues on my end, all tests passed. Hope it's just a github action quirk indeed.

I added some tests as requested and moved the file where it is supposed to be :smile: 

Thx

File: src/utils/random.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+function xfnv1a(str: string) {
```
Comment (harrisiirak) [inline src/utils/random.ts:1]: A very small thing. Let's also add JSDoc to describe purpose of all these functions.

**Review by harrisiirak — APPROVED** [review]
LGTM. Added 2 documentation related comments that would be nice resolve, but not blockers.

It's going to be part of `v5.2.0`. I'll merge it when I'm ready with other changes for this release. Hopefully at the end of this week or at the beginning of the next week.

**Comment by harrisiirak** [comment]
Just a heads-up that I haven't forgotten about this PR! Hopefully, I'll have more time this weekend to put together the release.
